### PR TITLE
[IMP] mail, mass_mailing, base_setup: always apply blacklist + option to bypass

### PR DESCRIPTION
**Pull request by pyduf**
Always apply the blacklist in mass_mail composition mode and add an option to
bypass it for example for transactional or incident emails.

Technical note:
- a special test with target model with partner_id has been added because
checking blacklist works slightly differently in that case: blacklisted email
are checked against recipients (instead of email_to).

Task-2834862

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/wizard/mail_compose_message.py [hunk 1]
```diff
@@ -512,17 +512,26 @@ def _process_state(self, mail_values_dict):
...
-                targets = self.env[self.model].browse(mail_values_dict.keys()).read(['email_normalized'])
-                # First extract email from recipient before comparing with blacklist
-                blacklisted_rec_ids.update(target['id'] for target in targets
-                                           if target['email_normalized'] in blacklist)
+                if issubclass(type(self.env[self.model]), self.pool['mail.thread.blacklist']):
+                    targets = self.env[self.model].browse(mail_values_dict.keys()).read(['email_normalized'])
+                    # First extract email from recipient before comparing with blacklist
+                    blacklisted_rec_ids.update(target['id'] for target in targets
+                                               if target['email_normalized'] in blacklist)
+                else:
```
Comment (std-odoo) [inline addons/mail/wizard/mail_compose_message.py:532]: please apply the stable comments here
(or even better, apply the change on the stable branch and then re-cherry pick the commit here)

File: addons/mail/wizard/mail_compose_message.py [hunk 2]
```diff
@@ -515,7 +515,7 @@ def _process_state(self, mail_values_dict):
     def _get_blacklist_record_ids(self, mail_values_dict, recipients_info):
         blacklisted_rec_ids = set()
         if self.composition_mode == 'mass_mail':
-            self.env['mail.blacklist'].flush(['email'])
+            self.env['mail.blacklist'].flush_model(['email', 'active'])
```
Comment (std-odoo) [inline addons/mail/wizard/mail_compose_message.py:523]: shouldn't be this in stable ?

File: addons/mail/wizard/mail_compose_message.py [hunk 3]
```diff
@@ -146,6 +146,11 @@ def default_get(self, fields):
         help='This option permanently removes any track of email after it\'s been sent, including from the Technical menu in the Settings, in order to preserve storage space of your Odoo database.')
     auto_delete_message = fields.Boolean('Delete Message Copy', help='Do not keep a copy of the email in the document communication history (mass mailing only)')
     mail_server_id = fields.Many2one('ir.mail_server', 'Outgoing mail server')
+    bypass_blacklist = fields.Boolean('Include Blacklist',
+                                      help='Include all recipients, even the blacklisted ones.'
+                                           'To use with caution and for non-marketing-related issues '
+                                           '(shortage of service, emergencies, …)'
+                                      )
```
Comment (std-odoo) [inline addons/mail/wizard/mail_compose_message.py:153]:
```suggestion
                                           '(shortage of service, emergencies, …)')
```

File: addons/test_mail/tests/test_performance.py [hunk 4]
```diff
@@ -967,7 +967,7 @@ def test_complex_tracking_template(self):
         rec1 = rec.with_context(active_test=False)      # to see inactive records
         self.assertEqual(rec1.message_partner_ids, self.partners | self.env.user.partner_id | self.user_portal.partner_id)
 
-        with self.assertQueryCount(__system__=20, emp=21):
+        with self.assertQueryCount(__system__=20, employee=21):
```
Comment (std-odoo) [inline addons/test_mail/tests/test_performance.py:970]: why this change ?

File: addons/mail/controllers/discuss.py [hunk 5]
```diff
@@ -432,6 +432,20 @@ def mail_thread_data(self, thread_model, thread_id, request_list, **kwargs):
         thread = request.env[thread_model].with_context(active_test=False).search([('id', '=', thread_id)])
         return thread._get_mail_thread_data(request_list)
 
+    def _mail_thread_messages_filter(self):
+        """ Allows to add AND clause on mail.message query to filter further the message returned in the chatter.
```
Comment (std-odoo) [inline addons/mail/controllers/discuss.py:436]: > Allows to add AND clause on mail.message query

Technically you can modify the domain as you want (not just with AND clause)

Comment (std-odoo) [inline addons/mail/controllers/discuss.py:436]: (just to follow the [pep](https://peps.python.org/pep-0257/))
```suggestion
        """Allows to add AND clause on mail.message query to filter further the message returned in the chatter.
```

File: addons/sms/controllers/discuss.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import odoo.addons.mail.controllers.discuss as discuss
+from odoo.osv.expression import OR
+
+
+class DiscussController(discuss.DiscussController):
+    def _mail_thread_messages_filter(self):
+        domain = super()._mail_thread_messages_filter()
+        # Exclude sms not sent because of a permanent failure: blacklisted, duplicated or opted out
+        domain.append(
```
Comment (std-odoo) [inline addons/sms/controllers/discuss.py:12]: I would use expression.AND instead

File: addons/sms/models/mail_message.py [hunk 7]
```diff
@@ -19,6 +19,7 @@ class MailMessage(models.Model):
     ], ondelete={'sms': lambda recs: recs.write({'message_type': 'email'})})
     has_sms_error = fields.Boolean(
         'Has SMS error', compute='_compute_has_sms_error', search='_search_has_sms_error')
+    sms_ids = fields.One2many('sms.sms', 'mail_message_id', string='Messages', groups="base.group_system")
```
Comment (std-odoo) [inline addons/sms/models/mail_message.py:22]: picky std
```suggestion
    sms_ids = fields.One2many('sms.sms', 'mail_message_id', string='Messages', groups='base.group_system')
```

File: addons/sms/wizard/sms_composer.py [hunk 8]
```diff
@@ -237,12 +237,16 @@ def _action_send_sms_mass(self, records=None):
...
-        sms_all = self._prepare_mass_sms(records, sms_record_values)
 
-        if sms_all and self.mass_keep_log and records and issubclass(type(records), self.pool['mail.thread']):
+        if sms_record_values and self.mass_keep_log and records and issubclass(type(records), self.pool['mail.thread']):
             log_values = self._prepare_mass_log_values(records, sms_record_values)
-            records._message_log_batch(**log_values)
+            mail_messages = records._message_log_batch(**log_values)
+            for mail_message in mail_messages:
+                sms_record_vals = sms_record_values.get(mail_message.res_id)
+                if sms_record_vals:
```
Comment (std-odoo) [inline addons/sms/wizard/sms_composer.py:246]: is it possible that this condition is False ?

**Review by std-odoo — COMMENTED** [review]
@pyduf Hi :) 

First batch of comments, good work overall :)

File: addons/mail/wizard/mail_compose_message.py (hunk 2 above)
Comment (pyduf) [reply addons/mail/wizard/mail_compose_message.py:523]: Actually, it something in stable (not in 15.0 but between 15.0 and master), it is just that the fix has changed it and this is just to restore it. Sorry about that.

File: addons/test_mail/tests/test_performance.py (hunk 4 above)
Comment (pyduf) [reply addons/test_mail/tests/test_performance.py:970]: That was an error (user is employee and not emp). And without that change, I get a query count error.

File: addons/mail/controllers/discuss.py (hunk 5 above)
Comment (pyduf) [reply addons/mail/controllers/discuss.py]: I have reworded it as "Allows to refine mail.message query..." to make it more generic but actually the methods append condition in a "AND" block. That was what I wanted to say but it was not clear enough.

Comment (pyduf) [reply addons/mail/controllers/discuss.py]: Changed, thanks.

File: addons/sms/controllers/discuss.py (hunk 6 above)
Comment (pyduf) [reply addons/sms/controllers/discuss.py]: Maybe I'm missing something but to use AND, I have to do it like this:
```python
return [AND([
            *super()._mail_thread_messages_filter(),
            # Exclude sms not sent because of a permanent failure: blacklisted, duplicated or opted out
            OR([
                [('sms_ids', '=', False)],  # Not all mail.message have a related sms.sms
                [('sms_ids.failure_type', 'not in', ('sms_blacklist', 'sms_optout', 'sms_duplicate'))],
            ])
        ])]
```
But the AND will anyway be applied by the class using the result of the method so it is a bit redundant (and anyway using an other operator like this would be dangerous as we don't know the order of execution of the inherited method).

I have updated the code with a shorter version but less efficient so depending on your answer I may come back to the previous one or another solution. Thanks

File: addons/sms/wizard/sms_composer.py (hunk 8 above)
Comment (pyduf) [reply addons/sms/wizard/sms_composer.py]: Yes I come to the same conclusion so I have updated the code.
Note that everything could be simplified in one line but I have kept the 2 operations: _prepare_mass_log_values and _message_log_batch in separate lines to make things clearer (using 2 variables only used once).

File: addons/mail/wizard/mail_compose_message.py [hunk 9]
```diff
@@ -146,6 +146,10 @@ def default_get(self, fields):
         help='This option permanently removes any track of email after it\'s been sent, including from the Technical menu in the Settings, in order to preserve storage space of your Odoo database.')
     auto_delete_message = fields.Boolean('Delete Message Copy', help='Do not keep a copy of the email in the document communication history (mass mailing only)')
     mail_server_id = fields.Many2one('ir.mail_server', 'Outgoing mail server')
+    bypass_blacklist = fields.Boolean('Include Blacklist',
+                                      help='Include all recipients, even the blacklisted ones.'
```
Comment (std-odoo) [inline addons/mail/wizard/mail_compose_message.py:150]: (same below)
```suggestion
                                      help='Include all recipients, even the blacklisted ones. '
```

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 10]
```diff
@@ -362,6 +362,8 @@
                                     <group string="Advanced" groups="base.group_no_one">
                                         <field name="mail_server_available" invisible="1"/>
                                         <field name="name" required="False" string="Name" attrs="{'readonly': [('state', 'in', ('sending', 'done'))]}"/>
+                                        <field name="bypass_blacklist"
+                                               attrs="{'readonly': [('state', 'not in', ('draft'))]}"/>
```
Comment (std-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:366]: does this work ?
(it will be interpreted as `['d', 'r', 'a', 'f', 't']` because `('draft')` is not a tuple but a string)
```suggestion
                                               attrs="{'readonly': [('state', 'not in', ('draft', ))]}"/>
```

File: addons/sms/wizard/sms_composer.py [hunk 11]
```diff
@@ -47,7 +47,11 @@ def default_get(self, fields):
     # options for comment and mass mode
     mass_keep_log = fields.Boolean('Keep a note on document', default=True)
     mass_force_send = fields.Boolean('Send directly', default=False)
-    mass_use_blacklist = fields.Boolean('Use blacklist', default=True)
+    mass_bypass_blacklist = fields.Boolean('Include Blacklist',
+                                           help='Include all recipients, even the blacklisted ones.'
```
Comment (std-odoo) [inline addons/sms/wizard/sms_composer.py:51]:
```suggestion
                                           help='Include all recipients, even the blacklisted ones. '
```

File: addons/mail/controllers/discuss.py [hunk 12]
```diff
@@ -443,10 +444,22 @@ def mail_thread_data(self, thread_model, thread_id, request_list, **kwargs):
 
     @http.route('/mail/thread/messages', methods=['POST'], type='json', auth='user')
     def mail_thread_messages(self, thread_model, thread_id, max_id=None, min_id=None, limit=30, **kwargs):
+        # Need a sudo because mail.mail has different ACL than mail.message
+        filtered_ids = request.env['mail.message'].sudo().search(
+            AND([
```
Comment (std-odoo) [inline addons/mail/controllers/discuss.py:463]: you can simplify it a bit like you did with `_mail_thread_messages_filter`, as you prefer

**Review by std-odoo — COMMENTED** [review]
@pyduf Hi :) 

Some last comments, then we should be good :)

File: addons/mail/wizard/mail_compose_message.py (hunk 9 above)
Comment (pyduf) [reply addons/mail/wizard/mail_compose_message.py]: Thanks, modified in the 3 places

File: addons/mass_mailing/views/mailing_mailing_views.xml (hunk 10 above)
Comment (pyduf) [reply addons/mass_mailing/views/mailing_mailing_views.xml]: Indeed, thanks, I have changed it. (but strangely, it works also without the comma...)

File: addons/mail/controllers/discuss.py (hunk 12 above)
Comment (pyduf) [reply addons/mail/controllers/discuss.py:465]: As discussed, now the method _mail_thread_messages_filter returns a domain and I use operator on them (so this one cannot be simplified).

File: addons/mail/wizard/mail_compose_message.py [hunk 13]
```diff
@@ -169,6 +169,10 @@ def _partner_ids_domain(self):
         'Keep Message Copy',
         compute="_compute_auto_delete_keep_log", readonly=False, store=True,
         help='Keep a copy of the email content if emails are removed (mass mailing only)')
+    bypass_blacklist = fields.Boolean('Include Blacklist',
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:172]: ``include_excluded``? That way if we go for a positive meaning (include excluded emails), we align field name with its meaning.

Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:172]: Would say something like "Include excluded emails" maybe ?

File: addons/mass_mailing/models/mailing.py [hunk 14]
```diff
@@ -1048,7 +1052,12 @@ def _get_view_url(self, email_to, res_id):
         return url
 
     def action_send_mail(self, res_ids=None):
-        author_id = self.env.user.partner_id.id
+        current_partner_id = self.env.user.partner_id.id
+        if self.env['ir.model.data']._xmlid_to_res_id('base.partner_root') == current_partner_id:
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:1056]: Actually, shouldn't we always consider the author is the "context user" computed here below (aka: mailing responsible or last writer or current user) ? Seems we could align everything.

File: addons/sms/wizard/sms_composer.py [hunk 15]
```diff
@@ -47,7 +47,11 @@ def default_get(self, fields):
     # options for comment and mass mode
     mass_keep_log = fields.Boolean('Keep a note on document', default=True)
     mass_force_send = fields.Boolean('Send directly', default=False)
-    mass_use_blacklist = fields.Boolean('Use blacklist', default=True)
+    mass_bypass_blacklist = fields.Boolean('Include Blacklist',
```
Comment (tde-banana-odoo) [inline addons/sms/wizard/sms_composer.py:50]: Note: to align with final name chosen :)

File: addons/mass_mailing/models/mailing.py [hunk 16]
```diff
@@ -173,6 +173,10 @@ def _get_default_mail_server_id(self):
         domain="[('mailing_model_name', '=', mailing_model_name)]")
     mailing_filter_domain = fields.Char('Favorite filter domain', related='mailing_filter_id.mailing_domain')
     mailing_filter_count = fields.Integer('# Favorite Filters', compute='_compute_mailing_filter_count')
+    bypass_blacklist = fields.Boolean('Include Blacklist',
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:176]: Note: to align with final name chosen :)

**Review by tde-banana-odoo — COMMENTED** [review]
Some comments while reading the diff, waiting decision about the new fields on mail.message before going further :)

**Comment by pyduf** [comment]
This PR is replaced with a simplified version: odoo/odoo#187403
